File: 42-trapping-rain-water/42-trapping-rain-water.py
class Solution:
    def trap(self, height: List[int]) -> int:
        left = right = []
        mx = height[0]
        for i in range(len(height)):
            mx = max(mx,height[i])
            left.append(mx)
        mx = height[-1]
        for i in range(len(height)-1,-1,-1):
            mx = max(mx,height[i])
            right.append(mx)
        right = right[::-1]
        water = 0
        for i in range(len(height)):
            water += abs(min(left[i],right[i])-height[i])
        return water
        

File: 42-trapping-rain-water/42-trapping-rain-water.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        int N = height.size();
        vector<int> left(N,0),right(N,0);
        int mx = height[0];
        for(int i = 0;i<N;i++){
            mx = max(mx,height[i]);
            left[i] = mx;
        }
        mx = height[N-1];
        for(int i = N-1;i>=0;i--){
            mx = max(mx,height[i]);
            right[i] = mx;
        }
        // reverse(right.begin(), right.end());
        int water = 0;
        for(int i = 0;i<N;i++){
            water += abs(min(left[i],right[i])-height[i]);
        }
        return water;
    }
};